### PR TITLE
Add Ray's scope to training arguments

### LINKED ISSUE #16683: Option to change Ray's gridsearch scope
# 🚀 Feature request

It would be great if we can get more control over how Ray is selecting the best trial in a hyperparameter search. Currently, the default value for [`get_best_trial`](https://docs.ray.io/en/latest/tune/api_docs/analysis.html#ray.tune.ExperimentAnalysis.get_best_trial) is used here:

https://github.com/huggingface/transformers/blob/7c5d79912a21880ce13d77881940458e90d98917/src/transformers/integrations.py#L299

namely `scope="last"`. So for each trial, it will simply take the _last_ checkpoint of that trial, and use its performance to compare with the other trials. This is not always ideal as it may very well be that some trials converge sooner than other and than overfit, leading to poor evaluation scores in their last checkpoints. Fortunately, Ray allows other options, such as `"all"`, which takes the best checkpoint instead of the last, of each trial and compares those.

## Motivation

Currently there is no way to pass this through to Ray from within `transformers` as I can tell, yet it is an important aspect of hyperparameter search.

## Your contribution

I can work on this if requested, although I am still looking for input how to best tackle this. One could simply add an argument to TrainingArguments, e.g. `ray_scope`, and then change this line

https://github.com/huggingface/transformers/blob/7c5d79912a21880ce13d77881940458e90d98917/src/transformers/integrations.py#L299

to

```python
best_trial = analysis.get_best_trial(metric="objective", mode=direction[:3], scope=trainer.args.ray_scope)
```

### PR DESCRIPTION
**Pull request by BramVanroy**
# What does this PR do?

As discussed in https://github.com/huggingface/transformers/issues/16683, it often happens that you want more control over gridsearch. By default, Ray will run different trials with different hyperparameters, and then select the best one based on the loss (or other metric) of each trial's _final_ checkpoint. That's not always what you want however (when some trials converge faster than others and therefore overfit in the same number of steps). Luckily Ray allows [other options](https://docs.ray.io/en/latest/tune/api_docs/analysis.html#ray.tune.ExperimentAnalysis.get_best_trial). With this PR, the user gets a little bit more control over how to use Ray's search.

Fixes #16683 

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. **Discussed but never received an official reply. Got some community responses of others who would like to see this change, however.**
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).

## Who can review?

@richardliaw, @amogkam

### CODE REVIEW

**Comment by richardliaw** [comment]
Awesome stuff :) @Yard1 could you help shepherd this PR?

**Comment by Yard1** [comment]
@BramVanroy Hey this looks great! Looks like you only need to run black to reformat the files.

**Comment by BramVanroy** [comment]
@richardliaw @Yard1 I can't manage to satisfy the quality checks (`doc-builder`). The issue being the very long Ray URL. Any thoughts how I should deal with this?

**Comment by Yard1** [comment]
In general, you can break up very long links with `\`, eg.
```
        ray_scope (`str`, *optional*, defaults to `"last"`):
            The scope to use when doing hyperparameter search with Ray. By default, `"last"` will be used. Ray will
            then use the last checkpoint of all trials, compare those, and select the best one. However, other options
            are also available. See the
            [Ray documentation](
            https://docs.ray.io/en/latest/tune/api_docs/analysis.html\
#ray.tune.ExperimentAnalysis.get_best_trial)
            for more options.
```

File: src/transformers/training_args.py [hunk 1]
```diff
@@ -456,6 +456,11 @@ class TrainingArguments:
         torchdynamo (`str`, *optional*):
             The token that is used to set the backend compiler for TorchDynamo. Possible choices are ["eager",
             "nvfuser]. This is an experimental API and subject to change.
+        ray_scope (`str`, *optional*, defaults to `"last"`):
+            The scope to use when doing hyperparameter search with Ray. By default, `"last"` will be used. Ray will
+            then use the last checkpoint of all trials, compare those, and select the best one. However, other options
+            are also available. See the [Ray documentation]( https://docs.ray.io/en/latest/tune/api_docs/analysis.html\
+            #ray.tune.ExperimentAnalysis.get_best_trial) for more options.
```
Comment (Yard1) [inline src/transformers/training_args.py:463]:
```suggestion
            are also available. See the [Ray documentation]( https://docs.ray.io/en/latest/tune/api_docs/analysis.html\
#ray.tune.ExperimentAnalysis.get_best_trial) for more options.
```

**Comment by LysandreJik** [comment]
cc @sgugger for the additional argument to the trainings arguments

**Review by sgugger — APPROVED** [review]
Thanks for your work!

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -456,6 +456,11 @@ class TrainingArguments:
         torchdynamo (`str`, *optional*):
             The token that is used to set the backend compiler for TorchDynamo. Possible choices are ["eager",
             "nvfuser]. This is an experimental API and subject to change.
+        ray_scope (`str`, *optional*, defaults to `"last"`):
+            The scope to use when doing hyperparameter search with Ray. By default, `"last"` will be used. Ray will
+            then use the last checkpoint of all trials, compare those, and select the best one. However, other options
+            are also available. See the [Ray documentation]( https://docs.ray.io/en/latest/tune/api_docs/analysis.html\
+#ray.tune.ExperimentAnalysis.get_best_trial) for more options.
```
[the remark this answers is not in the document]
Comment (Yard1) [reply src/transformers/training_args.py:463]: This formating is intentional to satisfy line length limits, see https://stackoverflow.com/questions/14414878/how-do-i-break-a-link-in-a-rst-docstring-to-satisfy-pep8 for example

[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/training_args.py:463]: Yes but we do not use rst docstrings, so this won't be rendered in our doc. Just put everything back in one line and `make style` will automatically format it the best it can be.

[the remark this answers is not in the document]
Comment (Yard1) [reply src/transformers/training_args.py:463]: cc @BramVanroy
